File: 20190528/53_Maximum_Subarray.cpp
class Solution {
public:
    int maxSubArray(vector<int>& nums) {
        int sum = 0;
        int res = nums[0];
        for (int i = 0; i < nums.size(); i++)
        {
            sum += nums[i];
            res = max(sum, res);
            if (sum < 0)
                sum = 0;
        }
        return res;      
        
        /****** O(N**2) ******/   
//         int res = nums[0];
    
//         for (int i = 0; i < nums.size(); i++)
//         {
//             int sum = 0;
            
//             for (int j = i ; j < nums.size(); j++)
//             {
//                 sum += nums[j];
//                 res = max(sum, res);
//             }
//         }
        
//         return res;
    }
};

File: 20190528/53_Maximum_Subarray.py
class Solution:
    def maxSubArray(self, nums: List[int]) -> int:
        res = nums[0]
        sum = 0
        for i in nums:
            sum += i
            res = max(res, sum)
            if sum < 0:
                sum = 0
        return res
            
        